### PR TITLE
Add failing test for multiple ranks and threads

### PR DESCRIPTION
**Pull request by jh66637**
@kronbichler @peterrum 

`MatrixFree::reinit()` fails with multiple ranks and threading enabled. I traced this down to the update of values on inner faces. The error reads 
```
An error occurred in line <1005> of file </home/johannes/workspace/dealii/include/deal.II/lac/dynamic_sparsity_pattern.h> in function
    void dealii::DynamicSparsityPattern::add(dealii::DynamicSparsityPattern::size_type, dealii::DynamicSparsityPattern::size_type)
The violated condition was: 
    ::dealii::deal_II_exceptions::internals::compare_less_than(i, n_rows())
Additional information: 
    Index 16 is not in the half-open range [0,16).
```
Using only one thread (`Utilities::MPI::MPI_InitFinalize mpi_init(argc, argv, 1)`) works as expected. Running with one rank and threading enabled works as well.

### CODE REVIEW

File: tests/matrix_free/reinit_matrix_free_threading.cc [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+int
+main(int argc, char **argv)
+{
+  // Init with threading enabled
+  Utilities::MPI::MPI_InitFinalize mpi_init(argc, argv);
+  mpi_initlog();
+
+  do_test<2,double>(3,3);
+  do_test<3,double>(3,3);
```
Comment (kronbichler) [inline tests/matrix_free/reinit_matrix_free_threading.cc:77]: Does the 3D test need to be such big? I would hope that 2 refinements and a polynomial degree 2 should suffice (the 2D variant is ok).

**Review by kronbichler — APPROVED** [review]
This is a good test. I checked #16340 with it. I would suggest to run it only with 2 MPI rank because I think that exposes the bug already. We should try not to use more resources than necessary, because we do MPI + threads, causing a lot of load on the tester.

**Comment by kronbichler** [comment]
You need to run the indent script on this test.

**Review by kronbichler — APPROVED** [review]

**Comment by kronbichler** [comment]
I suggest with merging this test until #16340 is merged, to be sure things are fine.

**Comment by jh66637** [comment]
This is fine for me. Thank you for the quick fix :)

**Comment by kronbichler** [comment]
Could you please rebase this PR to pass the test?

**Pull request merged**